### ISSUE TITLE
question / suggestion about UserTrade amounts

### ISSUE THREAD

**Issue by npomfret** [knowm/XChange#2645]
@timmolter i think i've seen some inconsistencies in how the amounts are represented in the `UserTrade` class.

Firstly, am I correct in thinking that the `UserTrade.originalAmount` is supposed to be represented in the base currency?  I believe that is the case everywhere.

I think some exchanges return quantities net of fees, and some return gross quantities. I think it would be nice if we removed this difference in xchange.  

It might be clearest if we say that the `UserTrade.originalAmount` field represents the gross amount traded in the base ccy (perhaps even call it `grossAmountTradedInBaseCcy`).  And so if you wanted to say compute what your balance change should be you would subtract the fee from that (if the fee currency was the same).

You could add the following methods:

```java
  public BigDecimal netBalanceChangeInBaseCcy() {
    BigDecimal netBalanceChange = this.grossAmountTradedInBaseCcy;
    if(feeCurrency.equals(currencyPair.base)) {
      netBalanceChange = netBalanceChange.subtract(feeAmount);
    }
    return netBalanceChange;
  }
  
  public BigDecimal netBalanceChangeInCounterCcy() {
    BigDecimal netBalanceChange = this.grossAmountTradedInBaseCcy.multiply(this.price);
    if(feeCurrency.equals(currencyPair.counter)) {
      netBalanceChange = netBalanceChange.subtract(feeAmount);
    }
    return netBalanceChange;
  }
```

That would mean in some exchanges adding the fee to the amount.  

For the moment though, perhaps we could consider just renaming the field, and over time people will spot and fix the places where net amounts are being used.

**Comment by timmolter** [comment]
yes, the amount should be the base currency. It's for example how much BTC you bought/sold for whatever counter currency (USD, ETH). 

I think `InBaseCcy` should be left out because it's already pretty clear what `Amount` is. Ity's what you're buying/selling. The price is then in counter currency terms, and it's the price for some amount of something you want to buy. I think that's just a fundamental basic concept people need to understand. 

I never realized that some exchanges return quantities net of fees, and some return gross quantities.  What's `UserTrade.originalAmount` for again anyway? Isn't if for the case of limit orders and partial fills?

**Comment by npomfret** [comment]
So `UserTrade` extends `Trade` which has the field `originalAmount`.  Possibly the meaning is the amount of the original order, but if so there's no place for the executed amount to go.  Obviously for partially executed orders these values can be different.  Some exchanges don't give you the original amount, only the executed amount.

I think there's an argument for both `originalAmount` and `executedAmount` to be in `UserTrade`.  

If we add it in there's the question of what is the correct value thats currently being passed, is it the executed amount or original amount?  (Personally I think executedAmount is more important and suspect that's what the majority of implementations are assuming).

There's a similar argument for the price.  Limit orders have an _original price_ and an _average executed price_, but there's aren't modelled in the UserTrade class.  Again, I think the average executed price is the more important one here.

**Comment by walec51** [comment]
The name may be misleading but the Trade class should NOT have two fields with the amount  in any case. A Trade is a Trade - it holds the information of that WAS actually traded - not how much the user intended to trade - you have the Order class . An order has originalAmount and cumulativeAmount because many Trades are done to fill one Order and they do not have to sell all of the originalAmount the user wanted to sell.

**Comment by timmolter** [comment]
OK, I looked at the code. I don't recall why `originalAmount` in `Trade` was changed from `amount` anymore. It should probably be switched back, and then the class left as is. 

The original function of `Trade` was for the `Trades` class when fetching executed trades from market data endpoints. It's correctly in the `marketdata` package.

`UserTrade` is in the `trade` package and adds these three fields to `Trade`:

```java
  /** The id of the order responsible for execution of this trade */
  private final String orderId;

  /** The fee that was charged by the exchange for this trade. */
  private final BigDecimal feeAmount;

  /** The currency in which the fee was charged. */
  private final Currency feeCurrency;
```
So, what you're saying is that for `UserTrade`s sometimes the amount is given net of fees and other times without fees? The simplest solution would be to define `amount`, via JavaDocs, as the net amount. Then in adapter code, modules need to work out the net amount and fee amount whether it gets returned as net or gross. Does that make sense?
